### PR TITLE
Add CFI support to Power architecture

### PR DESCRIPTION
**Pull request by tmcgilchrist**
Adds DWARF CFI information for the Power backend, which allows tools like debuggers and profilers to work on that platform.  I've tested on the Ubuntu 24.04 ppc64le platform I've access to, it should work on ppc64 big endian but I don't have access to a test machine for that. 

This brings backtraces and debugging support to the last native backend that was missing them.

### CODE REVIEW

**Comment by OlivierNicole** [comment]
I can review in absence of someone more relevant, but I lack skills on POWER, and will have to get up to speed regarding CFI. @xavierleroy Is there any chance that you have time to review this?

File: asmcomp/power/emit.mlp [hunk 1]
```diff
@@ -597,18 +597,21 @@ let emit_instr env i =
...
+        let n = frame_size env in
+        if env.f.fun_frame_required then begin
+            `	mflr	0\n`;
+            `	std	0, {emit_int 16}(1)\n`;
+            `	std	2, {emit_int 8}(1)\n`
+         end;
+        (* Always emit CFI for LR at CFA + 16. On Power ABI, the caller saves
+           LR at [caller_SP + 16] before making a call, so LR is always at
+           CFA + 16 whether we save it ourselves or not. This ensures the
+           debugger can find the return address for unwinding. *)
```
Comment (xavierleroy) [inline asmcomp/power/emit.mlp]: I'm afraid this comment is not quite right.  The caller reserves one word at position 16 in its stack frame, in which the callee can save LR if it needs to.  The caller never writes anything in this word.  If the callee is a leaf function, it doesn't need to save LR, and it seems incorrect to me to tell the debugger that LR's value is in that memory word: LR's value is still in the LR register.

File: asmcomp/power/emit.mlp [hunk 2]
```diff
@@ -597,18 +597,21 @@ let emit_instr env i =
...
+        cfi_offset ~reg: 65 (* LR *) ~offset: 16;
+        if n > 0 then begin
+            `	stdu	1, {emit_int(-n)}(1)\n`;
```
Comment (xavierleroy) [inline asmcomp/power/emit.mlp]: The original code was using `addi` to decrement the stack pointer.  It's correct to use `stdu` instead, and C compilers do generate `stdu` because it implements a "frame pointer" (a linked list of stack frames) for free.  Did you change to `stdu` in preparation for your "frame pointer support" PR (which I haven't read yet)?  Or for some other reason?

File: asmcomp/power/emit.mlp [hunk 3]
```diff
@@ -597,18 +597,21 @@ let emit_instr env i =
...
-        `	mflr	0\n`;
-        `	std	0, {emit_int ra}(1)\n`;
-        cfi_offset ~reg: 65 (* LR *) ~offset: (ra - n);
-        `	std	2, {emit_int(toc_save_offset env)}(1)\n`
-      end
+        let n = frame_size env in
+        if env.f.fun_frame_required then begin
+            `	mflr	0\n`;
+            `	std	0, {emit_int 16}(1)\n`;
+            `	std	2, {emit_int 8}(1)\n`
```
Comment (xavierleroy) [inline asmcomp/power/emit.mlp]: These 3 instructions (which save the LR and R2 registers in the reserved words of the caller's frame) used to be done after decrementing the stack pointer, but now they are done before.  I think it makes no difference.  Any reason for the change?

**Review by xavierleroy — COMMENTED** [review]
Thank you for giving this Power port some care.  I skimmed the PR and I trust you (and thank you!) for all the CFI magic you've added.  I only have a few comments on the code generated for function prologues, see below.

File: asmcomp/power/emit.mlp (hunk 2 above)
Comment (tmcgilchrist) [reply asmcomp/power/emit.mlp]: Yes, there is a frame pointer PR (https://github.com/ocaml/ocaml/pull/14505)  ready to go after this PR. 

I'll switch back to `addi` instruction since it is clearly not required for CFI, and use `stdu` only for frame pointers.

File: asmcomp/power/emit.mlp (hunk 3 above)
Comment (tmcgilchrist) [reply asmcomp/power/emit.mlp]: No functional reason.  I copied the ordering from the Power ELF V2 ABI document (Figure 2.19) and from GCC assembly output. 

```
; Without backchain:
mfr r0                      - Copy LR to R0
std r0, 16(r1)           - Store the LR in the previous LR save double word
addi r1, r1, -32        - Decrement SP
```

Comment (xavierleroy) [reply asmcomp/power/emit.mlp]: OK, thanks for the clarification.

File: asmcomp/power/emit.mlp (hunk 2 above)
Comment (xavierleroy) [reply asmcomp/power/emit.mlp]: Thanks for the clarification.  You can leave `stdu` if you want, but, yes, `addi` is enough for this PR.

File: asmcomp/power/emit.mlp [hunk 4]
```diff
@@ -597,18 +597,21 @@ let emit_instr env i =
...
+         end;
+        (* Always emit CFI for LR at CFA + 16. On Power ABI, the caller saves
+           LR at [caller_SP + 16] before making a call, so LR is always at
+           CFA + 16 whether we save it ourselves or not. This ensures the
+           debugger can find the return address for unwinding. *)
+        cfi_offset ~reg: 65 (* LR *) ~offset: 16;
```
Comment (xavierleroy) [inline asmcomp/power/emit.mlp]: I still believe the comment is misleading, and the `cfi_offset` should only be emitted if we've saved LR in the stack just before.  What about:
```suggestion
            cfi_offset ~reg: 65 (* LR *) ~offset: 16;
         end;       
```
I wasn't able to test this corner case, since small tail functions don't have a `Lprologue` instruction to begin with, so no LR saving and no `cfi_offset` takes place.  Still, I think the suggestion above is more logical.

Comment (tmcgilchrist) [reply asmcomp/power/emit.mlp]: Yes that makes sense. Thank you.

**Comment by hhugo** [comment]
> it should work on ppc64 big endian but I don't have access to a test machine for that.

You should be able to request access to https://portal.cfarm.net/, they have 2 ppc64 big endian machines.

**Comment by tmcgilchrist** [comment]
@hhugo I've applied for an account, now waiting for it to be approved. I'll post test results from there once I have access to a power64be machine.

Fixed one remaining TSAN issue, where I missed adding CFI_STARTPROC/CFI_ENDPROC to caml_tsan_exit_on_raise_asm. The regular `--enable-cfi` option passes all tests, `--enable-tsan` fails on some tests but that needs more work in https://github.com/ocaml/ocaml/pull/14302. Perhaps that could be done in a follow up PR once this is merged?

These tests fail with `--enable-cfi --enable-tsan` before and after my changes in this PR. 
```
List of failed tests:                                                                    
    tests/lib-bigarray-file/mapfile.ml                                                   
    tests/lib-channels/close_during_flush.ml                                             
    tests/lib-channels/input_all.ml                                                      
    tests/lib-runtime-events/test.ml                                                     
    tests/lib-runtime-events/test_caml.ml
    tests/lib-runtime-events/test_caml_counters.ml
    tests/lib-runtime-events/test_caml_exception.ml
    tests/lib-runtime-events/test_caml_parallel.ml
    tests/lib-runtime-events/test_caml_reentry.ml
    tests/lib-runtime-events/test_caml_runparams.ml
    tests/lib-runtime-events/test_caml_slot_reuse.ml
    tests/lib-runtime-events/test_caml_stubs_gc.ml
    tests/lib-runtime-events/test_compact.ml 
    tests/lib-runtime-events/test_corrupted.ml
    tests/lib-runtime-events/test_create_cursor_failures.ml
    tests/lib-runtime-events/test_dropped_events.ml
    tests/lib-runtime-events/test_env_start.ml
    tests/lib-runtime-events/test_external.ml
    tests/lib-runtime-events/test_external_preserve.ml
    tests/lib-runtime-events/test_fork.ml
    tests/lib-runtime-events/test_instrumented.ml
    tests/lib-runtime-events/test_user_event.ml
    tests/lib-runtime-events/test_user_event_unknown.ml
    tests/lib-scanf-2/tscanf2_master.ml
    tests/lib-threads/fileio.ml
    tests/lib-unix/common/channel_of.ml
    tests/lib-unix/common/cloexec.ml
    tests/lib-unix/common/process_pid.ml
    tests/lib-unix/common/redirections.ml
    tests/lib-unix/common/test_unix_cmdline.ml
    tests/lib-unix/common/truncate.ml
    tests/lib-unix/common/wait_nohang.ml
    tests/parallel/backup_thread_pipe.ml
    tests/parallel/multicore_systhreads.ml
    tests/parallel/recommended_domain_count_unix.ml
    tests/statmemprof/bigarray.ml
    tests/tsan/exn_from_c_stack_args.ml
```

This is ready to go, if you are happy with the code @xavierleroy

**Comment by tmcgilchrist** [comment]
> You should be able to request access to https://portal.cfarm.net/, they have 2 ppc64 big endian machines.

Unfortunately they're both ELFv1 ABI and we only support ELFv2 ABI on power for the native backend (https://github.com/ocaml/ocaml/issues/12656). Nice POWER10 hardware though.

**Comment by OlivierNicole** [comment]
Reading this discussion, my understanding is that @xavierleroy's comments have been addressed and this is only missing his formal approval.

**Comment by tmcgilchrist** [comment]
I'm rebasing this change ~~and testing a corner case I missed originally around `caml_runstack`. Please hold off until I can test the fix.~~ (edit: the issue was a false alarm).

**Comment by tmcgilchrist** [comment]
This should be good to go @OlivierNicole @xavierleroy

**Comment by OlivierNicole** [comment]
There seems to be many test failures on the macOS-x86_64 CI, do you know what could cause them?

**Comment by OlivierNicole** [comment]
NVM, I re-ran the job and it clearly was a CI fluke.

**Comment by OlivierNicole** [comment]
I discussed this with @xavierleroy and this is good to merge after a final round of CI. Regarding the assembly, we trust that your changes are correct, and in addition they visibly don’t change the semantics outside of DWARF.

I took the liberty to force-push to remove a minor conflict in the changelog.
